### PR TITLE
Changes to default construction and base-class semantic analysis

### PR DESCRIPTION
**Pull request by KevinEady**
### Removed DefaultConstructorFunction

As discussed, removed the compiler-generated `DefaultConstructorFunction`. This required some additional semantic analysis changes:
  - A class that does not define a constructor cannot be called as a constructor
  - A base class must define a constructor

### Changed `ScopableName::operator<` 

Changed operator. such that scopable names `<nullopt>::Foo` and `Foo::Foo` are equal. This allows a function call's function link `Foo()` (with no scope specified) to match a registered function `Foo::Foo`. 

The previous implementation incorrectly added multiple `AvailableParseTree`s for the same user function, causing multiple AST passes. I've added the `internal_error()` call to catch this.

### CODE REVIEW

File: pol-core/bscript/compiler/analyzer/SemanticAnalyzer.cpp [hunk 1]
```diff
@@ -144,24 +144,56 @@ void SemanticAnalyzer::visit_block( Block& block )
...
 }
 
+void SemanticAnalyzer::visit_class_parameter_declaration( ClassParameterDeclaration& node )
+{
+  auto constructor = node.constructor_link->function();
+  if ( !constructor )
+  {
+    auto msg = fmt::format( "Base class '{}' does not define a constructor.", node.name );
+
+    auto func_itr =
```
Comment (KevinEady) [inline pol-core/bscript/compiler/analyzer/SemanticAnalyzer.cpp:154]: comment whats happening

File: pol-core/bscript/compiler/astbuilder/UserFunctionBuilder.cpp [hunk 2]
```diff
@@ -47,22 +48,29 @@ std::unique_ptr<ClassDeclaration> UserFunctionBuilder::class_declaration(
...
       for ( auto parameter_name : param_list->IDENTIFIER() )
       {
-        parameters.push_back( std::make_unique<Identifier>( location_for( *parameter_name ),
-                                                            text( parameter_name ) ) );
+        auto baseclass_name = text( parameter_name );
+
+        auto class_param_decl = std::make_unique<ClassParameterDeclaration>(
+            location_for( *parameter_name ), baseclass_name );
+
+        workspace.function_resolver.register_function_link(
```
Comment (KevinEady) [inline pol-core/bscript/compiler/astbuilder/UserFunctionBuilder.cpp:64]: doc

File: pol-core/bscript/compiler/astbuilder/UserFunctionVisitor.cpp [hunk 3]
```diff
@@ -28,8 +27,6 @@ antlrcpp::Any UserFunctionVisitor::visitFunctionDeclaration(
 
   auto uf = tree_builder.function_declaration( ctx, scope );
   workspace.function_resolver.register_user_function( scope, uf.get() );
-  if ( uf->type == UserFunctionType::Constructor )
```
Comment (KevinEady) [inline pol-core/bscript/compiler/astbuilder/UserFunctionVisitor.cpp:31]: Was removed due to `ScopableName::operator<` change

File: pol-core/bscript/compiler/model/FunctionLink.h [hunk 4]
```diff
@@ -28,6 +29,10 @@ class FunctionLink
   // eg. `Animal` when calling any function inside the `Animal` class
   const std::string calling_scope;
 
+  // If a FunctionLink can only be linked to a constructor. Used inside
+  // UserFunctionBuilder::visit_class_declaration
```
Comment (KevinEady) [inline pol-core/bscript/compiler/model/FunctionLink.h:33]: ... and `ClassParameterDeclaration` to link to the constructor of a class

File: pol-core/bscript/compiler/optimizer/ReferencedFunctionGatherer.cpp [hunk 5]
```diff
@@ -40,6 +41,11 @@ void ReferencedFunctionGatherer::visit_function_expression( FunctionExpression&
   reference( *fr.function_link );
 }
 
+void ReferencedFunctionGatherer::visit_class_parameter_declaration( ClassParameterDeclaration& cpd )
+{
+  reference( *cpd.constructor_link );
```
Comment (KevinEady) [inline pol-core/bscript/compiler/optimizer/ReferencedFunctionGatherer.cpp:46]: why

File: testsuite/escript/classes/fail-multiple-base.src
Comment (KevinEady) [inline testsuite/escript/classes/fail-multiple-base.src:1]: missing .err

File: testsuite/escript/classes/fail-self-base.src [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// A class can have a variable the same as the classname, but that cannot be
```
Comment (KevinEady) [inline testsuite/escript/classes/fail-self-base.src:1]: wrong comment

File: testsuite/escript/classes/scopes-var-as-classname.src [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// A class can have a variable the same as the classname. We should not error on
+// a call, but we should error when used as a base-class (test TODO)
```
Comment (KevinEady) [inline testsuite/escript/classes/scopes-var-as-classname.src:2]: can implement a test for this todo

**Review by turleypol — APPROVED** [review]
If you want to bring back default constructor is up to you :)

**Pull request merged**